presto-verifier/src/main/java/com/facebook/presto/verifier/framework/DataVerification.java

```diff
@@ -91,23 +103,54 @@ public DataMatchResult verify(
             ChecksumQueryContext controlChecksumQueryContext,
             ChecksumQueryContext testChecksumQueryContext)
     {
-        List<Column> controlColumns = getColumns(getHelperAction(), typeManager, control.getObjectName());
         List<Column> testColumns = getColumns(getHelperAction(), typeManager, test.getObjectName());
-
-        Query controlChecksumQuery = checksumValidator.generateChecksumQuery(control.getObjectName(), controlColumns);
         Query testChecksumQuery = checksumValidator.generateChecksumQuery(test.getObjectName(), testColumns);
-
-        controlChecksumQueryContext.setChecksumQuery(formatSql(controlChecksumQuery));
         testChecksumQueryContext.setChecksumQuery(formatSql(testChecksumQuery));
 
-        QueryResult<ChecksumResult> controlChecksum = callAndConsume(
-                () -> getHelperAction().execute(controlChecksumQuery, CONTROL_CHECKSUM, ChecksumResult::fromResultSet),
-                stats -> stats.getQueryStats().map(QueryStats::getQueryId).ifPresent(controlChecksumQueryContext::setChecksumQueryId));
+        List<Column> controlColumns = null;
+        ChecksumResult controlChecksumResult = null;
+
+        if (!skipControl || saveSnapshot) {
+            controlColumns = getColumns(getHelperAction(), typeManager, control.getObjectName());
+            Query controlChecksumQuery = checksumValidator.generateChecksumQuery(control.getObjectName(), controlColumns);
+            controlChecksumQueryContext.setChecksumQuery(formatSql(controlChecksumQuery));
+
+            QueryResult<ChecksumResult> controlChecksum = callAndConsume(
+                    () -> getHelperAction().execute(controlChecksumQuery, CONTROL_CHECKSUM, ChecksumResult::fromResultSet),
+                    stats -> stats.getQueryStats().map(QueryStats::getQueryId).ifPresent(controlChecksumQueryContext::setChecksumQueryId));
+            controlChecksumResult = getOnlyElement(controlChecksum.getResults());
+
+            if (saveSnapshot) {
+                String snapshot = ChecksumResult.toJson(Optional.of(controlChecksumResult));
```

Comment:
curious why we are wrapping checksum result in an `Optional` and inside `toJson` we are calling `.get` without any `isPresent` checks. Can we get rid of this `Optional` wrapping since it looks like not needed.

Reply:
Changed. Thanks